node/mock_test.go

```diff
@@ -18,11 +17,60 @@ var (
 	_ PodNotifier         = (*mockProvider)(nil)
 )
 
+type waitableInt struct {
+	cond *sync.Cond
+	val  int
+}
+
+func newWaitableInt() *waitableInt {
+	return &waitableInt{
+		cond: sync.NewCond(&sync.Mutex{}),
+	}
+}
+
+func (w *waitableInt) read() int {
+	defer w.cond.L.Unlock()
+	w.cond.L.Lock()
+	return w.val
+}
+
+func (w *waitableInt) until(ctx context.Context, f func(int) bool) error {
+	c := make(chan struct{})
+	defer close(c)
+	go func() {
+		select {
+		case <-ctx.Done():
+			w.cond.Broadcast()
+		case <-c:
+		}
+	}()
+
+	for {
```

Comment:
```go
w.cond.L.Lock()
defer w.cond.L.Unlock()

for !f(w.val) {
    if err := ctx.Err() != nil {
        return err
    }
    w.cond.Wait()
}
return nil
```

Reply:
Won't that deadlock on itself? because when w.cond.Wait returns, it’ll lock w.cond.L. and on the 2nd loop, it’ll lock twice.

Reply:
Wait unlocks when it's called and re-locks when it returns.
Here's the official example:

>  Wait atomically unlocks c.L and suspends execution of the calling goroutine. After later resuming execution, Wait locks c.L before returning. Unlike in other systems, Wait cannot return unless awoken by Broadcast or Signal.
>
> Because c.L is not locked when Wait first resumes, the caller typically cannot assume that the condition is true when Wait returns. Instead, the caller should Wait in a loop: 
```
c.L.Lock()
for !condition() {
    c.Wait()
}
... make use of condition ...
c.L.Unlock()
```

Reply:
nevermind